File: server/app/utils/data_quality_monitor.py
"""数据质量监控(Bug-130)
完整性/准确性/一致性/及时性/唯一性规则 + 异常告警 + 统计面板
"""

from __future__ import annotations

import re
import threading
import time
from collections import deque
from dataclasses import dataclass, field
from datetime import datetime
from enum import StrEnum
from typing import Any


class DQDimension(StrEnum):
    COMPLETENESS = "COMPLETENESS"  # 完整性
    ACCURACY = "ACCURACY"  # 准确性
    CONSISTENCY = "CONSISTENCY"  # 一致性
    TIMELINESS = "TIMELINESS"  # 及时性
    UNIQUENESS = "UNIQUENESS"  # 唯一性
    VALIDITY = "VALIDITY"  # 有效性


class AlertLevel(StrEnum):
    OK = "OK"
    INFO = "INFO"
    WARN = "WARN"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


@dataclass
class DQRule:
    rule_id: str
    name: str
    dimension: DQDimension
    dataset: str
    target_field: str = ""
    threshold: float = 0.0
    comparison: str = "lt"  # lt/le/gt/ge/eq
    params: dict[str, Any] = field(default_factory=dict)
    enabled: bool = True
    description: str = ""
    created_at: float = field(default_factory=time.time)


@dataclass
class DQSample:
    sample_id: str
    dataset: str
    data: dict[str, Any]
    received_at: float = field(default_factory=time.time)
    source: str = ""


@dataclass
class DQViolation:
    rule_id: str
    dataset: str
    field: str
    dimension: DQDimension
    level: AlertLevel
    actual_value: float
    threshold: float
    sample_id: str = ""
    message: str = ""
    detected_at: float = field(default_factory=time.time)  # type: ignore[operator]


@dataclass
class DQMetric:
    rule_id: str
    dataset: str
    dimension: DQDimension
    total: int
    failed: int
    rate: float
    window_start: float
    window_end: float

    def status(self, threshold: float, comparison: str) -> AlertLevel:
        if comparison == "lt":
            if self.rate < threshold:
                return AlertLevel.CRITICAL
        elif comparison == "le":
            if self.rate <= threshold:
                return AlertLevel.CRITICAL
        elif comparison == "gt":
            if self.rate > threshold:
                return AlertLevel.CRITICAL
        elif comparison == "ge" and self.rate >= threshold:
            return AlertLevel.CRITICAL
        return AlertLevel.OK if self.failed == 0 else AlertLevel.WARN


class _Status:
    OK = "OK"


@dataclass
class DQConfig:
    window_sec: float = 300.0
    max_samples: int = 10000
    max_violations: int = 5000
    max_alerts_per_rule: int = 50


def _compare(value: float, threshold: float, comparison: str) -> bool:
    if comparison == "lt":
        return value < threshold
    if comparison == "le":
        return value <= threshold
    if comparison == "gt":
        return value > threshold
    if comparison == "ge":
        return value >= threshold
    if comparison == "eq":
        return value == threshold
    return False


class DataQualityMonitor:
    """数据质量监控"""

    def __init__(self, config: DQConfig | None = None) -> None:
        self.config = config or DQConfig()
        self._rules: dict[str, DQRule] = {}
        self._samples: deque[DQSample] = deque(maxlen=self.config.max_samples)
        self._violations: deque[DQViolation] = deque(maxlen=self.config.max_violations)
        self._alerts_by_rule: dict[str, int] = {}
        self._seen_keys: dict[str, float] = {}  # 唯一性
        self._lock = threading.RLock()

    def add_rule(self, rule: DQRule) -> None:
        with self._lock:
            self._rules[rule.rule_id] = rule

    def remove_rule(self, rule_id: str) -> bool:
        with self._lock:
            return self._rules.pop(rule_id, None) is not None

    def enable_rule(self, rule_id: str, enabled: bool = True) -> bool:
        with self._lock:
            r = self._rules.get(rule_id)
            if r is None:
                return False
            r.enabled = enabled
            return True

    def get_rule(self, rule_id: str) -> DQRule | None:
        return self._rules.get(rule_id)

    def list_rules(self, dataset: str | None = None) -> list[DQRule]:
        with self._lock:
            rs = list(self._rules.values())
        if dataset:
            rs = [r for r in rs if r.dataset == dataset]
        return rs

    def _generate_sample_id(self) -> str:
        return f"s-{int(time.time() * 1000)}-{len(self._samples)}"

    def _check_completeness(self, rule: DQRule, data: dict[str, Any]) -> bool:
        field = rule.target_field
        if field not in data:
            return False
        v = data[field]
        if v is None:
            return False
        if isinstance(v, str) and v.strip() == "":
            return False
        return not (isinstance(v, (list, dict)) and len(v) == 0)

    def _check_accuracy(self, rule: DQRule, data: dict[str, Any]) -> bool:
        field = rule.target_field
        if field not in data:
            return False
        v = data[field]
        if v is None:
            return False
        allowed = rule.params.get("allowed_values")
        if allowed is not None and v not in allowed:
            return False
        pattern = rule.params.get("pattern")
        if pattern is not None and isinstance(v, str):
            import re

            if not re.match(pattern, v):
                return False
        lo = rule.params.get("min_value")
        hi = rule.params.get("max_value")
        if isinstance(v, (int, float)) and not isinstance(v, bool):
            if lo is not None and v < lo:
                return False
            if hi is not None and v > hi:
                return False
        return True

    def _check_consistency(self, rule: DQRule, data: dict[str, Any]) -> bool:
        pairs = rule.params.get("field_pairs", [])
        for pair in pairs:
            a, b = pair.get("a"), pair.get("b")
            if a not in data or b not in data:
                return False
            if data[a] != data[b]:
                return False
        formula = rule.params.get("formula")
        if formula is not None:
            # 支持 sum_greater_than(a,b,c) > value
            try:
                fname, rest = formula.split("(", 1)
                args_str, comp = rest.rsplit(")", 1)
                args = [a.strip() for a in args_str.split(",")]
                _, val = comp[0], comp[1:].strip()
                vals = [data.get(a) for a in args]
                if any(v is None for v in vals):
                    return False
                target = float(val)
                if fname == "sum_greater_than":
                    if not (sum(vals) > target):  # type: ignore[arg-type]
                        return False
                elif fname == "sum_less_than":
                    if not (sum(vals) < target):  # type: ignore[arg-type]
                        return False
                elif fname == "diff_equals" and abs(vals[0] - vals[1]) != target:  # type: ignore[operator]
                    return False
            except Exception:
                return False
        return True

    def _check_timeliness(self, rule: DQRule, data: dict[str, Any], received_at: float) -> bool:
        field = rule.target_field
        if field not in data:
            return False
        ts = data[field]
        if not isinstance(ts, (int, float)):
            try:
                from datetime import datetime

                ts = datetime.fromisoformat(str(ts).replace("Z", "")).timestamp()
            except Exception:
                return False
        max_delay = float(rule.params.get("max_delay_sec", 0))
        return (received_at - float(ts)) <= max_delay

    def _check_uniqueness(self, rule: DQRule, data: dict[str, Any], sample_id: str) -> bool:
        field = rule.target_field
        if field not in data:
            return False
        v = data[field]
        if v is None or v == "":
            return True  # 唯一性只对非空值有意义
        fingerprint = f"{rule.dataset}:{field}:{v}"
        with self._lock:
            if fingerprint in self._seen_keys:
                return False
            self._seen_keys[fingerprint] = time.time()
        return True

    def _check_validity(self, rule: DQRule, data: dict[str, Any]) -> bool:
        field = rule.target_field
        if field not in data:
            return False
        v = data[field]
        if v is None:
            return False
        validator = rule.params.get("validator")
        if validator == "email":
            return bool(re.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$", str(v)))
        if validator == "phone":
            return bool(re.match(r"^1[3-9]\d{9}$", str(v)))
        if validator == "uuid":
            return bool(
                re.match(r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$", str(v))
            )
        if validator == "non_negative":
            return isinstance(v, (int, float)) and not isinstance(v, bool) and v >= 0
        if validator == "non_empty":
            if isinstance(v, str):
                return v.strip() != ""
            if isinstance(v, (list, dict)):
                return len(v) > 0
            return True
        return True

    def _dispatch_check(self, rule: DQRule, data: dict[str, Any], sample_id: str, received_at: float) -> bool:
        if rule.dimension == DQDimension.COMPLETENESS:
            return self._check_completeness(rule, data)
        if rule.dimension == DQDimension.ACCURACY:
            return self._check_accuracy(rule, data)
        if rule.dimension == DQDimension.CONSISTENCY:
            return self._check_consistency(rule, data)
        if rule.dimension == DQDimension.TIMELINESS:
            return self._check_timeliness(rule, data, received_at)
        if rule.dimension == DQDimension.UNIQUENESS:
            return self._check_uniqueness(rule, data, sample_id)
        if rule.dimension == DQDimension.VALIDITY:
            return self._check_validity(rule, data)
        return True

    def feed(self, dataset: str, data: dict[str, Any], source: str = "") -> DQSample:
        sample = DQSample(
            sample_id=self._generate_sample_id(),
            dataset=dataset,
            data=dict(data),
            source=source,
            received_at=time.time(),
        )
        with self._lock:
            self._samples.append(sample)
            for rule in self._rules.values():
                if not rule.enabled or rule.dataset != dataset:
                    continue
                passed = self._dispatch_check(rule, data, sample.sample_id, sample.received_at)
                if not passed:
                    self._record_violation(rule, sample.sample_id, data)
        return sample

    def _record_violation(self, rule: DQRule, sample_id: str, data: dict[str, Any]) -> None:
        actual = 0.0
        if rule.dimension in (
            DQDimension.COMPLETENESS,
            DQDimension.ACCURACY,
            DQDimension.CONSISTENCY,
            DQDimension.VALIDITY,
        ):
            actual = 0.0
        elif rule.dimension == DQDimension.TIMELINESS:
            field = rule.target_field
            ts = data.get(field)
            max_delay = float(rule.params.get("max_delay_sec", 0))
            try:
                if isinstance(ts, str):
                    ts = datetime.fromisoformat(ts.replace("Z", "")).timestamp()
                actual = time.time() - float(ts) - max_delay  # type: ignore[arg-type]
            except Exception:
                actual = 0.0
        elif rule.dimension == DQDimension.UNIQUENESS:
            actual = 1.0
        level = AlertLevel.ERROR
        if rule.dimension == DQDimension.TIMELINESS:
            level = AlertLevel.CRITICAL if actual > 0 else AlertLevel.WARN
        violation = DQViolation(
            rule_id=rule.rule_id,
            dataset=rule.dataset,
            field=rule.target_field,
            dimension=rule.dimension,
            level=level,
            actual_value=actual,
            threshold=rule.threshold,
            sample_id=sample_id,
            message=f"{rule.name} 检测到违规",
        )
        with self._lock:
            self._violations.append(violation)
            cnt = self._alerts_by_rule.get(rule.rule_id, 0)
            if cnt < self.config.max_alerts_per_rule:
                self._alerts_by_rule[rule.rule_id] = cnt + 1

    def get_violations(
        self, dataset: str | None = None, rule_id: str | None = None, limit: int = 100
    ) -> list[DQViolation]:
        with self._lock:
            vs = list(self._violations)
        if dataset:
            vs = [v for v in vs if v.dataset == dataset]
        if rule_id:
            vs = [v for v in vs if v.rule_id == rule_id]
        return vs[-limit:]

    def get_metrics(self, dataset: str | None = None, window_sec: float | None = None) -> list[DQMetric]:
        window = window_sec or self.config.window_sec
        now = time.time()
        start = now - window
        with self._lock:
            samples = [s for s in self._samples if s.received_at >= start and (dataset is None or s.dataset == dataset)]
        metrics: list[DQMetric] = []
        for rule in self._rules.values():
            if dataset and rule.dataset != dataset:
                continue
            if not rule.enabled:
                continue
            total = 0
            failed = 0
            for s in samples:
                if s.dataset != rule.dataset:
                    continue
                total += 1
                if not self._dispatch_check(rule, s.data, s.sample_id, s.received_at):
                    failed += 1
            rate = (total - failed) / total if total > 0 else 1.0
            metrics.append(
                DQMetric(
                    rule_id=rule.rule_id,
                    dataset=rule.dataset,
                    dimension=rule.dimension,
                    total=total,
                    failed=failed,
                    rate=rate,
                    window_start=start,
                    window_end=now,
                )
            )
        return metrics

    def health(self, dataset: str | None = None) -> dict[str, Any]:
        metrics = self.get_metrics(dataset=dataset)
        bad = []
        for m in metrics:
            rule = next((r for r in self._rules.values() if r.rule_id == m.rule_id), None)
            threshold = rule.threshold if rule is not None else 0
            comparison = rule.comparison if rule is not None else "lt"
            if _compare(m.rate, threshold, comparison):
                bad.append(m)
        return {
            "ok": len(bad) == 0,
            "bad_metrics": [{"rule_id": m.rule_id, "rate": m.rate, "total": m.total, "failed": m.failed} for m in bad],
            "metric_count": len(metrics),
        }

    def stats(self) -> dict[str, Any]:
        with self._lock:
            return {
                "rules_total": len(self._rules),
                "rules_enabled": sum(1 for r in self._rules.values() if r.enabled),
                "samples_total": len(self._samples),
                "violations_total": len(self._violations),
                "unique_keys_tracked": len(self._seen_keys),
            }

    def clear(self) -> None:
        with self._lock:
            self._samples.clear()
            self._violations.clear()
            self._alerts_by_rule.clear()
            self._seen_keys.clear()

File: apps/api/src/utils/data-quality-monitor.ts
/**
 * 数据质量监控.
 *
 * 完整性 / 准确性 / 一致性 / 及时性 / 唯一性 / 有效性 规则
 * + 异常告警 + 统计面板.
 *
 * 参考: git show 3ee96cf0:server/app/utils/data_quality_monitor.py
 */

/** 数据质量维度. */
export enum DQDimension {
  COMPLETENESS = 'COMPLETENESS',
  ACCURACY = 'ACCURACY',
  CONSISTENCY = 'CONSISTENCY',
  TIMELINESS = 'TIMELINESS',
  UNIQUENESS = 'UNIQUENESS',
  VALIDITY = 'VALIDITY',
}

/** 告警级别. */
export enum AlertLevel {
  OK = 'OK',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR',
  CRITICAL = 'CRITICAL',
}

/** 比较运算符. */
export type Comparison = 'lt' | 'le' | 'gt' | 'ge' | 'eq'

/** 数据质量规则. */
export interface DQRule {
  ruleId: string
  name: string
  dimension: DQDimension
  dataset: string
  targetField: string
  threshold: number
  comparison: Comparison
  params: Record<string, unknown>
  enabled: boolean
  description: string
  createdAt: number
}

/** 数据样本. */
export interface DQSample {
  sampleId: string
  dataset: string
  data: Record<string, unknown>
  receivedAt: number
  source: string
}

/** 违规记录. */
export interface DQViolation {
  ruleId: string
  dataset: string
  field: string
  dimension: DQDimension
  level: AlertLevel
  actualValue: number
  threshold: number
  sampleId: string
  message: string
  detectedAt: number
}

/** 单规则指标. */
export interface DQMetric {
  ruleId: string
  dataset: string
  dimension: DQDimension
  total: number
  failed: number
  rate: number
  windowStart: number
  windowEnd: number
}

/** 配置. */
export interface DQConfig {
  windowSec: number
  maxSamples: number
  maxViolations: number
  maxAlertsPerRule: number
}

export const DEFAULT_DQ_CONFIG: DQConfig = {
  windowSec: 300,
  maxSamples: 10_000,
  maxViolations: 5000,
  maxAlertsPerRule: 50,
}

/** 比较运算. */
function compare(value: number, threshold: number, comp: Comparison): boolean {
  switch (comp) {
    case 'lt':
      return value < threshold
    case 'le':
      return value <= threshold
    case 'gt':
      return value > threshold
    case 'ge':
      return value >= threshold
    case 'eq':
      return value === threshold
  }
}

/** 创建规则的便捷工厂. */
export function createRule(
  opts: Partial<DQRule> & Pick<DQRule, 'ruleId' | 'dimension' | 'dataset'>,
): DQRule {
  return {
    name: opts.name ?? opts.ruleId,
    targetField: opts.targetField ?? '',
    threshold: opts.threshold ?? 0,
    comparison: opts.comparison ?? 'lt',
    params: opts.params ?? {},
    enabled: opts.enabled ?? true,
    description: opts.description ?? '',
    createdAt: opts.createdAt ?? Date.now() / 1000,
    ...opts,
  } as DQRule
}

/**
 * 数据质量监控器.
 */
export class DataQualityMonitor {
  private readonly config: DQConfig
  private readonly rules = new Map<string, DQRule>()
  private readonly samples: DQSample[] = []
  private readonly violations: DQViolation[] = []
  private readonly alertsByRule = new Map<string, number>()
  /** 唯一性: fingerprint -> ts */
  private readonly seenKeys = new Map<string, number>()
  private sampleCounter = 0

  constructor(config: Partial<DQConfig> = {}) {
    this.config = { ...DEFAULT_DQ_CONFIG, ...config }
  }

  /** 添加规则. */
  addRule(rule: DQRule): void {
    this.rules.set(rule.ruleId, rule)
  }

  /** 移除规则. */
  removeRule(ruleId: string): boolean {
    return this.rules.delete(ruleId)
  }

  /** 启用/禁用规则. */
  enableRule(ruleId: string, enabled = true): boolean {
    const r = this.rules.get(ruleId)
    if (!r) return false
    r.enabled = enabled
    return true
  }

  /** 获取规则. */
  getRule(ruleId: string): DQRule | undefined {
    return this.rules.get(ruleId)
  }

  /** 列出规则 (可按 dataset 过滤). */
  listRules(dataset?: string): DQRule[] {
    const rs = Array.from(this.rules.values())
    return dataset ? rs.filter((r) => r.dataset === dataset) : rs
  }

  private generateSampleId(): string {
    this.sampleCounter += 1
    return `s-${Math.floor(Date.now())}-${this.sampleCounter}`
  }

  /** 完整性: 字段存在且非空. */
  private checkCompleteness(rule: DQRule, data: Record<string, unknown>): boolean {
    const field = rule.targetField
    if (!(field in data)) return false
    const v = data[field]
    if (v === null || v === undefined) return false
    if (typeof v === 'string' && v.trim() === '') return false
    if (Array.isArray(v) && v.length === 0) return false
    if (typeof v === 'object' && v !== null && !Array.isArray(v) && Object.keys(v).length === 0)
      return false
    return true
  }

  /** 准确性: 枚举值 / 正则 / 数值范围. */
  private checkAccuracy(rule: DQRule, data: Record<string, unknown>): boolean {
    const field = rule.targetField
    if (!(field in data)) return false
    const v = data[field]
    if (v === null || v === undefined) return false
    const allowed = rule.params['allowedValues'] as unknown[] | undefined
    if (allowed && !allowed.includes(v)) return false
    const pattern = rule.params['pattern'] as string | undefined
    if (pattern && typeof v === 'string') {
      if (!new RegExp(pattern).test(v)) return false
    }
    const lo = rule.params['minValue'] as number | undefined
    const hi = rule.params['maxValue'] as number | undefined
    if (typeof v === 'number' && !Number.isNaN(v)) {
      if (lo !== undefined && v < lo) return false
      if (hi !== undefined && v > hi) return false
    }
    return true
  }

  /** 一致性: 字段对相等 / 简单公式. */
  private checkConsistency(rule: DQRule, data: Record<string, unknown>): boolean {
    const pairs = (rule.params['fieldPairs'] as Array<{ a: string; b: string }> | undefined) ?? []
    for (const pair of pairs) {
      if (!(pair.a in data) || !(pair.b in data)) return false
      if (data[pair.a] !== data[pair.b]) return false
    }
    return true
  }

  /** 及时性: 时间戳字段在允许延迟内. */
  private checkTimeliness(
    rule: DQRule,
    data: Record<string, unknown>,
    receivedAt: number,
  ): boolean {
    const field = rule.targetField
    if (!(field in data)) return false
    const raw = data[field]
    let ts: number
    if (typeof raw === 'number') {
      ts = raw
    } else if (typeof raw === 'string') {
      const parsed = Date.parse(raw)
      if (Number.isNaN(parsed)) return false
      ts = parsed / 1000
    } else {
      return false
    }
    const maxDelay = Number(rule.params['maxDelaySec'] ?? 0)
    return receivedAt - ts <= maxDelay
  }

  /** 唯一性: 字段值在 dataset 内不重复. */
  private checkUniqueness(rule: DQRule, data: Record<string, unknown>): boolean {
    const field = rule.targetField
    if (!(field in data)) return false
    const v = data[field]
    if (v === null || v === '' || v === undefined) return true
    const fp = `${rule.dataset}:${field}:${String(v)}`
    if (this.seenKeys.has(fp)) return false
    this.seenKeys.set(fp, Date.now() / 1000)
    return true
  }

  /** 有效性: email/phone/uuid/非负/非空. */
  private checkValidity(rule: DQRule, data: Record<string, unknown>): boolean {
    const field = rule.targetField
    if (!(field in data)) return false
    const v = data[field]
    if (v === null || v === undefined) return false
    const validator = rule.params['validator'] as string | undefined
    switch (validator) {
      case 'email':
        return /^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(String(v))
      case 'phone':
        return /^1[3-9]\d{9}$/.test(String(v))
      case 'uuid':
        return /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/.test(
          String(v),
        )
      case 'non_negative':
        return typeof v === 'number' && !Number.isNaN(v) && v >= 0
      case 'non_empty':
        if (typeof v === 'string') return v.trim() !== ''
        if (Array.isArray(v) || (typeof v === 'object' && v !== null))
          return Object.keys(v).length > 0
        return true
      default:
        return true
    }
  }

  /** 分派到对应维度的检查. */
  private dispatchCheck(rule: DQRule, data: Record<string, unknown>, receivedAt: number): boolean {
    switch (rule.dimension) {
      case DQDimension.COMPLETENESS:
        return this.checkCompleteness(rule, data)
      case DQDimension.ACCURACY:
        return this.checkAccuracy(rule, data)
      case DQDimension.CONSISTENCY:
        return this.checkConsistency(rule, data)
      case DQDimension.TIMELINESS:
        return this.checkTimeliness(rule, data, receivedAt)
      case DQDimension.UNIQUENESS:
        return this.checkUniqueness(rule, data)
      case DQDimension.VALIDITY:
        return this.checkValidity(rule, data)
      default:
        return true
    }
  }

  /** 喂入一条样本, 触发规则检查. */
  feed(dataset: string, data: Record<string, unknown>, source = ''): DQSample {
    const sample: DQSample = {
      sampleId: this.generateSampleId(),
      dataset,
      data: { ...data },
      source,
      receivedAt: Date.now() / 1000,
    }
    this.samples.push(sample)
    if (this.samples.length > this.config.maxSamples) this.samples.shift()
    for (const rule of this.rules.values()) {
      if (!rule.enabled || rule.dataset !== dataset) continue
      const passed = this.dispatchCheck(rule, data, sample.receivedAt)
      if (!passed) this.recordViolation(rule, sample.sampleId, data)
    }
    return sample
  }

  private recordViolation(rule: DQRule, sampleId: string, data: Record<string, unknown>): void {
    let actual = 0
    let level = AlertLevel.ERROR
    if (rule.dimension === DQDimension.TIMELINESS) {
      const raw = data[rule.targetField]
      let ts = 0
      if (typeof raw === 'number') ts = raw
      else if (typeof raw === 'string') ts = Date.parse(raw) / 1000
      const maxDelay = Number(rule.params['maxDelaySec'] ?? 0)
      actual = Date.now() / 1000 - ts - maxDelay
      level = actual > 0 ? AlertLevel.CRITICAL : AlertLevel.WARN
    } else if (rule.dimension === DQDimension.UNIQUENESS) {
      actual = 1
    }
    const violation: DQViolation = {
      ruleId: rule.ruleId,
      dataset: rule.dataset,
      field: rule.targetField,
      dimension: rule.dimension,
      level,
      actualValue: actual,
      threshold: rule.threshold,
      sampleId,
      message: `${rule.name} 检测到违规`,
      detectedAt: Date.now() / 1000,
    }
    this.violations.push(violation)
    if (this.violations.length > this.config.maxViolations) this.violations.shift()
    const cnt = this.alertsByRule.get(rule.ruleId) ?? 0
    if (cnt < this.config.maxAlertsPerRule) this.alertsByRule.set(rule.ruleId, cnt + 1)
  }

  /** 获取违规列表. */
  getViolations(opts: { dataset?: string; ruleId?: string; limit?: number } = {}): DQViolation[] {
    const { dataset, ruleId, limit = 100 } = opts
    let vs = this.violations
    if (dataset) vs = vs.filter((v) => v.dataset === dataset)
    if (ruleId) vs = vs.filter((v) => v.ruleId === ruleId)
    return vs.slice(-limit)
  }

  /** 计算窗口内指标. */
  getMetrics(opts: { dataset?: string; windowSec?: number } = {}): DQMetric[] {
    const window = opts.windowSec ?? this.config.windowSec
    const now = Date.now() / 1000
    const start = now - window
    const samples = this.samples.filter(
      (s) => s.receivedAt >= start && (opts.dataset === undefined || s.dataset === opts.dataset),
    )
    const metrics: DQMetric[] = []
    for (const rule of this.rules.values()) {
      if (opts.dataset && rule.dataset !== opts.dataset) continue
      if (!rule.enabled) continue
      let total = 0
      let failed = 0
      for (const s of samples) {
        if (s.dataset !== rule.dataset) continue
        total += 1
        if (!this.dispatchCheck(rule, s.data, s.receivedAt)) failed += 1
      }
      const rate = total > 0 ? (total - failed) / total : 1
      metrics.push({
        ruleId: rule.ruleId,
        dataset: rule.dataset,
        dimension: rule.dimension,
        total,
        failed,
        rate,
        windowStart: start,
        windowEnd: now,
      })
    }
    return metrics
  }

  /** 健康检查. */
  health(dataset?: string): { ok: boolean; badMetrics: DQMetric[]; metricCount: number } {
    const metrics = this.getMetrics({ dataset })
    const bad = metrics.filter((m) => {
      const rule = this.rules.get(m.ruleId)
      if (!rule) return false
      return compare(m.rate, rule.threshold, rule.comparison)
    })
    return { ok: bad.length === 0, badMetrics: bad, metricCount: metrics.length }
  }

  /** 统计. */
  stats(): {
    rulesTotal: number
    rulesEnabled: number
    samplesTotal: number
    violationsTotal: number
    uniqueKeysTracked: number
  } {
    let rulesEnabled = 0
    for (const r of this.rules.values()) if (r.enabled) rulesEnabled += 1
    return {
      rulesTotal: this.rules.size,
      rulesEnabled,
      samplesTotal: this.samples.length,
      violationsTotal: this.violations.length,
      uniqueKeysTracked: this.seenKeys.size,
    }
  }

  /** 清空. */
  clear(): void {
    this.samples.length = 0
    this.violations.length = 0
    this.alertsByRule.clear()
    this.seenKeys.clear()
  }
}

/** 全局单例. */
export const dataQualityMonitor = new DataQualityMonitor()
